File: utils/simulator/power.py
"""
Power system models for mains-powered and solar direct drive (SDD) refrigerators.

Each model determines when the compressor can run and produces
power-related telemetry fields for the output records.
"""

import math
import random
import datetime as dt
from dataclasses import dataclass
from typing import Optional, Tuple

from utils.simulator.config import PowerConfig


@dataclass
class PowerState:
    """Mutable power system state carried between intervals.

    cumulative_powered_s: Total seconds with power available (ACCD or DCCD).
    battery_soc: State of charge for the logger battery (0.0 to 1.0).
        For SDD fridges this is the small backup battery that powers the
        data logger, NOT a battery for the compressor.
    in_outage: Whether a mains outage is currently active.
    outage_end: Timestamp when current outage ends (mains only).
    """
    cumulative_powered_s: float = 0.0
    battery_soc: float = 0.8
    in_outage: bool = False
    outage_end: Optional[dt.datetime] = None


class MainsPowerModel:
    """Mains power with stochastic outage events."""

    def __init__(self, config: PowerConfig, rng: random.Random):
        self.config = config
        self.rng = rng

    def _check_outage(self, state: PowerState, timestamp: dt.datetime, interval_s: float) -> PowerState:
        """Update outage state for the current interval."""
        if state.in_outage:
            if state.outage_end is not None and timestamp >= state.outage_end:
                state.in_outage = False
                state.outage_end = None
        else:
            # Probability of starting an outage in this interval
            hours = interval_s / 3600.0
            p = 1.0 - (1.0 - self.config.outage_probability_per_hour) ** hours
            if self.rng.random() < p:
                duration_h = self.rng.expovariate(1.0 / self.config.mean_outage_duration_hours)
                state.in_outage = True
                state.outage_end = timestamp + dt.timedelta(hours=duration_h)
        return state

    def simulate_interval(
        self,
        state: PowerState,
        timestamp: dt.datetime,
        interval_s: float,
        compressor_runtime_s: float,
    ) -> Tuple[PowerState, dict]:
        """Compute power readings for one interval.

        Returns:
            Tuple of (updated PowerState, dict with SVA, ACCD, ACSV).
        """
        state = self._check_outage(state, timestamp, interval_s)

        if state.in_outage:
            sva = 0
            acsv = 0.0
            powered_s = 0.0
        else:
            sva = self.config.nominal_voltage
            # Add small voltage variation
            acsv = round(sva + self.rng.gauss(0, sva * 0.02), 1)
            powered_s = interval_s

        state.cumulative_powered_s += powered_s

        readings = {
            'SVA': sva,
            'ACCD': round(powered_s, 1),
            'ACSV': round(acsv, 1),
        }
        return state, readings

    def is_power_available(self, state: PowerState) -> bool:
        return not state.in_outage


class SolarPowerModel:
    """Solar direct drive (SDD) power model.

    In an SDD fridge the compressor runs directly from solar panels — there
    is no electrical battery for the compressor.  Energy is stored thermally
    in the ice lining.  The BLOG field reflects a small primary-cell or
    rechargeable battery that keeps the *data logger* running when solar
    power is unavailable.
    """

    def __init__(self, config: PowerConfig, rng: random.Random):
        self.config = config
        self.rng = rng

    def _solar_voltage(self, timestamp: dt.datetime) -> float:
        """Compute instantaneous DC solar voltage from a bell curve."""
        hour = timestamp.hour + timestamp.minute / 60.0 + timestamp.second / 3600.0
        solar_noon = (self.config.sunrise_hour + self.config.sunset_hour) / 2.0
        daylight_hours = self.config.sunset_hour - self.config.sunrise_hour

        if hour < self.config.sunrise_hour or hour > self.config.sunset_hour:
            return 0.0

        phase = math.pi * (hour - solar_noon) / daylight_hours
        voltage = self.config.peak_dcsv * max(0.0, math.cos(phase))
        # Add noise for cloud variation
        noise = self.rng.gauss(0, voltage * 0.05) if voltage > 0 else 0
        return max(0.0, voltage + noise)

    def simulate_interval(
        self,
        state: PowerState,
        timestamp: dt.datetime,
        interval_s: float,
        compressor_runtime_s: float,
    ) -> Tuple[PowerState, dict]:
        """Compute power readings and update logger battery SOC.

        Returns:
            Tuple of (updated PowerState, dict with DCSV, DCCD, BLOG, BEMD).
        """
        dcsv = round(self._solar_voltage(timestamp), 1)
        solar_power_available = dcsv >= self.config.min_operating_voltage

        # Logger battery dynamics (BLOG)
        # The logger battery is a small cell that charges from solar and
        # slowly drains overnight to keep the logger recording data.
        hours = interval_s / 3600.0
        if solar_power_available:
            # Charge logger battery from solar (fast — small battery)
            charge_rate = self.config.charge_efficiency * 0.3  # SOC/hour
            state.battery_soc += charge_rate * hours
        else:
            # Logger drains battery slowly (very low power draw)
            drain_rate = 0.015  # SOC/hour — small enough to last overnight
            state.battery_soc -= drain_rate * hours

        state.battery_soc = max(0.0, min(1.0, state.battery_soc))

        # DCCD: cumulative seconds with DC power available to the compressor
        powered_s = interval_s if solar_power_available else 0.0
        state.cumulative_powered_s += powered_s

        # BLOG/BEMD: logger battery voltage mapped from SOC
        blog = round(self.config.blog_voltage_empty + state.battery_soc * self.config.blog_voltage_range, 1)

        readings = {
            'DCSV': dcsv,
            'DCCD': round(powered_s, 1),
            'BLOG': blog,
            'BEMD': blog,
        }
        return state, readings

    def is_power_available(self, state: PowerState, timestamp: dt.datetime) -> bool:
        """Compressor power is available only when solar voltage is sufficient."""
        dcsv = self._solar_voltage(timestamp)
        return dcsv >= self.config.min_operating_voltage

File: js/src/power.js
/**
 * Power system models for mains-powered and solar direct drive (SDD) refrigerators.
 *
 * Each model determines when the compressor can run and produces
 * power-related telemetry fields for the output records.
 */

import { PowerConfig } from "./config.js";

/**
 * Mutable power system state carried between intervals.
 */
export class PowerState {
  /**
   * @param {object} opts
   * @param {number} [opts.cumulative_powered_s=0.0] - Total seconds with power available.
   * @param {number} [opts.battery_soc=0.8] - Logger battery state of charge (0.0 to 1.0).
   * @param {boolean} [opts.in_outage=false] - Whether a mains outage is currently active.
   * @param {Date|null} [opts.outage_end=null] - Timestamp when current outage ends (mains only).
   */
  constructor({
    cumulative_powered_s = 0.0,
    battery_soc = 0.8,
    in_outage = false,
    outage_end = null,
  } = {}) {
    this.cumulative_powered_s = cumulative_powered_s;
    this.battery_soc = battery_soc;
    this.in_outage = in_outage;
    this.outage_end = outage_end;
  }
}

/**
 * Mains power with stochastic outage events.
 */
export class MainsPowerModel {
  /**
   * @param {PowerConfig} config
   * @param {import('./random.js').SeededRandom} rng
   */
  constructor(config, rng) {
    this.config = config;
    this.rng = rng;
  }

  /**
   * Update outage state for the current interval.
   * @param {PowerState} state
   * @param {Date} timestamp
   * @param {number} interval_s
   * @returns {PowerState}
   */
  _check_outage(state, timestamp, interval_s) {
    if (state.in_outage) {
      if (state.outage_end !== null && timestamp >= state.outage_end) {
        state.in_outage = false;
        state.outage_end = null;
      }
    } else {
      // Probability of starting an outage in this interval
      const hours = interval_s / 3600.0;
      const p =
        1.0 - Math.pow(1.0 - this.config.outage_probability_per_hour, hours);
      if (this.rng.random() < p) {
        // Exponential duration: Python's expovariate(1/mean) = mean * -ln(U)
        const u = this.rng.random();
        const duration_h =
          -Math.log(u) * this.config.mean_outage_duration_hours;
        state.in_outage = true;
        state.outage_end = new Date(
          timestamp.getTime() + duration_h * 3600 * 1000
        );
      }
    }
    return state;
  }

  /**
   * Compute power readings for one interval.
   *
   * power_available_override mirrors FaultEffects.power_available_override:
   * when false (e.g. a POWER_OUTAGE fault) the mains supply is forced off for
   * this interval even if no stochastic outage is active, so SVA/ACSV/ACCD all
   * reflect the outage. When null/undefined, behaviour is purely stochastic.
   *
   * @param {PowerState} state
   * @param {Date} timestamp
   * @param {number} interval_s
   * @param {number} compressor_runtime_s
   * @param {boolean|null} [power_available_override=null]
   * @returns {[PowerState, {SVA: number, ACCD: number, ACSV: number}]}
   */
  simulate_interval(
    state,
    timestamp,
    interval_s,
    compressor_runtime_s,
    power_available_override = null
  ) {
    state = this._check_outage(state, timestamp, interval_s);

    // Either a modeled stochastic outage or a fault that forces power off
    // leaves the appliance with no usable AC supply this interval.
    const outage = state.in_outage || power_available_override === false;

    let acsv, powered_s;
    if (outage) {
      acsv = 0.0;
      powered_s = 0.0;
    } else {
      const nominal = this.config.nominal_voltage;
      // ACSV is the average AC supply voltage; add small variation.
      acsv = round1(nominal + this.rng.gauss(0, nominal * 0.02));
      powered_s = interval_s;
    }

    state.cumulative_powered_s += powered_s;

    // SVA is the number of SECONDS within the (15-min) period that AC voltage
    // was in-bounds: the full interval when powered, 0 during an outage.
    // The interop schema bounds SVA to [0, 900].
    const sva = Math.round(Math.min(900.0, powered_s));

    // ACCD is the AC current (amps) drawn by the appliance, modeled from the
    // compressor duty cycle: a small always-on baseline plus the running
    // compressor draw scaled by the fraction of the interval it ran. With no
    // AC supply (mains outage) the appliance draws no current, so ACCD is 0 A
    // per the PQS E006 DS01.2 Annex-1 ACCD minimum of 0 (Data Format 00.00).
    const duty = interval_s > 0 ? compressor_runtime_s / interval_s : 0.0;
    let accd = outage
      ? 0.0
      : this.config.mains_baseline_current_a +
        this.config.mains_compressor_current_a * duty;
    accd = round2(Math.min(49.99, Math.max(0.0, accd)));

    const readings = {
      SVA: sva,
      ACCD: accd,
      ACSV: round1(acsv),
    };
    return [state, readings];
  }

  /**
   * @param {PowerState} state
   * @returns {boolean}
   */
  is_power_available(state) {
    return !state.in_outage;
  }
}

/**
 * Solar direct drive (SDD) power model.
 *
 * In an SDD fridge the compressor runs directly from solar panels -- there
 * is no electrical battery for the compressor. Energy is stored thermally
 * in the ice lining. The BLOG field reflects a small primary-cell or
 * rechargeable battery that keeps the data logger running when solar
 * power is unavailable.
 */
export class SolarPowerModel {
  /**
   * @param {PowerConfig} config
   * @param {import('./random.js').SeededRandom} rng
   */
  constructor(config, rng) {
    this.config = config;
    this.rng = rng;
  }

  /**
   * Compute instantaneous DC solar voltage from a bell curve.
   * @param {Date} timestamp
   * @returns {number}
   */
  _solar_voltage(timestamp) {
    const hour =
      timestamp.getUTCHours() +
      timestamp.getUTCMinutes() / 60.0 +
      timestamp.getUTCSeconds() / 3600.0;
    const solar_noon =
      (this.config.sunrise_hour + this.config.sunset_hour) / 2.0;
    const daylight_hours =
      this.config.sunset_hour - this.config.sunrise_hour;

    if (hour < this.config.sunrise_hour || hour > this.config.sunset_hour) {
      return 0.0;
    }

    const phase = (Math.PI * (hour - solar_noon)) / daylight_hours;
    const voltage = this.config.peak_dcsv * Math.max(0.0, Math.cos(phase));
    // Add noise for cloud variation
    const noise = voltage > 0 ? this.rng.gauss(0, voltage * 0.05) : 0;
    return Math.max(0.0, voltage + noise);
  }

  /**
   * Compute power readings and update logger battery SOC.
   * power_available_override mirrors FaultEffects.power_available_override:
   * when false (e.g. a POWER_OUTAGE fault) power is forced unavailable for this
   * interval regardless of solar voltage, so the compressor draws no DC current
   * and the logger battery drains.
   *
   * @param {PowerState} state
   * @param {Date} timestamp
   * @param {number} interval_s
   * @param {number} compressor_runtime_s
   * @param {boolean|null} [power_available_override=null]
   * @returns {[PowerState, {DCSV: number, DCCD: number, BLOG: number, BEMD: number}]}
   */
  simulate_interval(
    state,
    timestamp,
    interval_s,
    compressor_runtime_s,
    power_available_override = null
  ) {
    const dcsv = round1(this._solar_voltage(timestamp));
    let solar_power_available = dcsv >= this.config.min_operating_voltage;
    if (power_available_override === false) {
      solar_power_available = false;
    }

    // Logger battery dynamics (BLOG)
    const hours = interval_s / 3600.0;
    if (solar_power_available) {
      // Charge logger battery from solar (fast -- small battery)
      const charge_rate = this.config.charge_efficiency * 0.3; // SOC/hour
      state.battery_soc += charge_rate * hours;
    } else {
      // Logger drains battery slowly (very low power draw)
      const drain_rate = 0.015; // SOC/hour
      state.battery_soc -= drain_rate * hours;
    }

    state.battery_soc = Math.max(0.0, Math.min(1.0, state.battery_soc));

    const powered_s = solar_power_available ? interval_s : 0.0;
    state.cumulative_powered_s += powered_s;

    // DCCD is the DC current (amps) drawn by the compressor, modeled from the
    // compressor duty cycle while solar power is available (an SDD compressor
    // runs directly off the panels, so it draws no DC current without solar).
    // DCCD range is [0, 99.9]; the PQS E006 DS01.2 Annex-1 Data Format is
    // '00.0' (1 decimal place), unlike ACCD's '00.00'.
    const duty = interval_s > 0 ? compressor_runtime_s / interval_s : 0.0;
    let dccd = solar_power_available
      ? this.config.solar_compressor_current_a * duty
      : 0.0;
    dccd = round1(Math.min(99.9, Math.max(0.0, dccd)));

    // BLOG/BEMD: estimated DAYS of battery life remaining, scaled from the
    // logger battery state of charge (schema range [0, 9999.9]).
    const blog = round1(state.battery_soc * this.config.blog_full_days);
    const bemd = round1(state.battery_soc * this.config.bemd_full_days);

    const readings = {
      DCSV: dcsv,
      DCCD: dccd,
      BLOG: blog,
      BEMD: bemd,
    };
    return [state, readings];
  }

  /**
   * Compressor power is available only when solar voltage is sufficient.
   * @param {PowerState} state
   * @param {Date} timestamp
   * @returns {boolean}
   */
  is_power_available(state, timestamp) {
    const dcsv = this._solar_voltage(timestamp);
    return dcsv >= this.config.min_operating_voltage;
  }
}

/** Round to 1 decimal place. */
function round1(v) {
  return Math.round(v * 10) / 10;
}

/** Round to 2 decimal places. */
function round2(v) {
  return Math.round(v * 100) / 100;
}
